raw_linux.go

```diff
@@ -97,67 +91,55 @@ func newPacketConn(ifi *net.Interface, s socket, pbe uint16, sleeper sleeper) (*
 	})
 
 	return &packetConn{
-		ifi:     ifi,
-		s:       s,
-		sleeper: sleeper,
+		ifi: ifi,
+		s:   s,
 	}, err
 }
 
 // ReadFrom implements the net.PacketConn.ReadFrom method.
 func (p *packetConn) ReadFrom(b []byte) (int, net.Addr, error) {
-	// Set up deadline context if needed, if a read timeout is set
-	ctx, cancel := context.TODO(), func() {}
-	p.timeoutMu.RLock()
-	if p.rtimeout.After(time.Now()) {
-		ctx, cancel = context.WithDeadline(context.Background(), p.rtimeout)
+	var timeout time.Duration
+	var hasTimeout bool
+
+	p.timeoutMu.Lock()
+	if !p.rtimeout.IsZero() {
+		timeout = p.rtimeout.Sub(time.Now())
+		hasTimeout = true
+	}
+	p.timeoutMu.Unlock()
+
+	if hasTimeout && timeout < time.Microsecond {
+		return 0, nil, &timeoutError{}
+	}
+
+	tv := &unix.Timeval{
+		Sec:  int64(timeout / time.Second),
+		Usec: int64(timeout % time.Microsecond),
+	}
+
+	err := unix.SetsockoptTimeval(int(p.s.FD()), unix.SOL_SOCKET, unix.SO_RCVTIMEO, tv)
+	if err != nil {
+		return 0, nil, err
 	}
-	p.timeoutMu.RUnlock()
 
 	// Information returned by syscall.Recvfrom
 	var n int
 	var addr syscall.Sockaddr
-	var err error
 
 	for {
-		// Continue looping, or if deadline is set and has expired, return
-		// an error
-		select {
-		case <-ctx.Done():
-			// We only know how to handle deadline exceeded, so return any
-			// other errors for the caller to deal with
-			if err := ctx.Err(); err != context.DeadlineExceeded {
-				return n, nil, err
-			}
-
-			// Return standard net.OpError so caller can detect timeouts and retry
-			return n, nil, &net.OpError{
-				Op:   "read",
-				Net:  "raw",
-				Addr: nil,
-				Err:  &timeoutError{},
-			}
-		default:
-			// Not timed out, keep trying
-		}
 
 		// Attempt to receive on socket
 		n, addr, err = p.s.Recvfrom(b, 0)
 		if err != nil {
```

Comment:
On Linux at least, it looks like we need to check for `EAGAIN` here and return a `&timeoutError{}` too.

Reply:
The man pages say that `EAGAIN` only happens for non-blocking sockets. We have a blocking socket here.

Reply:
Interesting, when running this locally on Linux, I'm seeing:

```
$ time sudo ./arpc -d 2s -ip 192.168.1.100
2017/11/02 12:37:23 resource temporarily unavailable
sudo ./arpc -d 2s -ip 192.168.1.100  0.00s user 0.01s system 0% cpu 1.945 total
```

Which is just `EAGAIN` in string form.  I think that the setsockopt call might be setting non-blocking for you.

Reply:
I am going to check this later.

Reply:
On non-blocking sockets `EAGAIN` seems to indicate a timeout. I modified my changeset to return a `timeoutError` in that case.